ethereum/contracts/LightClientBridge.sol

```diff
@@ -0,0 +1,272 @@
+     * validator who has claimed to have signed the payload
+     * @param validatorPublicKey the public key of an arbitrary validator
+     * @param validatorSignatureCommitment the signature of an arbitrary validator
+     * @param validatorPublicKeyMerkleProof Proof required for validation of the Merkle tree
+     */
+    function newSignatureCommitment(
+        bytes32 payload,
+        uint256 validatorClaimsBitfield,
+        bytes memory validatorSignatureCommitment,
+        address validatorPublicKey,
+        bytes32[] calldata validatorPublicKeyMerkleProof
+    ) public payable {
+        /**
+         * @dev Check if validatorPublicKeyMerkleProof is valid based on ValidatorRegistry merkle root
+         */
+        require(
+            validatorRegistry.checkValidatorInSet(
+                validatorPublicKey,
+                validatorPublicKeyMerkleProof
+            ),
+            "Error: Sender must be in validator set"
+        );
+        // TODO also add a check here that the signing validator is in validatorClaimsBitfield
+        /**
+         * @dev Check if validatorSignatureCommitment is correct, ie. check if it matches
+         * the signature of senderPublicKey on the payload
+         */
+        require(
+            ECDSA.recover(payload, validatorSignatureCommitment) ==
+                validatorPublicKey,
+            "Error: Invalid Signature"
+        );
+        /**
+         * @dev Check that the bitfield actually contains enough claims to be succesful, ie, > 2/3
+         */
+        // TODO
+        /**
+         * @todo Lock up the sender stake as collateral
+         */
+        // TODO
+        // Accept and save the commitment
+        validationData[currentId] = ValidationData(
+            msg.sender,
+            payload,
+            validatorClaimsBitfield,
+            block.number
+        );
+        emit InitialVerificationSuccessful(msg.sender, block.number, currentId);
+        currentId = currentId.add(1);
+    }
+
+    /**
+     * @notice Performs the second step in the validation logic
+     * @param id an identifying value generated in the previous transaction
+     * @param payload contains the payload signed by the validator(s)
+     * @param randomSignatureCommitments an array of signatures from the randomly chosen validators
+     * @param randomSignatureBitfieldPositions an array of bitfields from the chosen validators
+     * @param randomValidatorAddresses an array of the ethereum address of each signer
+     * @param randomPublicKeyMerkleProofs an array of merkle proofs from the chosen validators
+     */
+    function completeSignatureCommitment(
+        uint256 id,
+        bytes32 payload,
+        bytes[] memory randomSignatureCommitments,
+        uint8[] memory randomSignatureBitfieldPositions,
+        address[] memory randomValidatorAddresses,
+        bytes32[][] memory randomPublicKeyMerkleProofs
+    ) public {
```

Comment:
This could potentially be quite a lot more efficient if we use `calldata` instead of memory. Especially because of the variable-length arrays we're passing in.

Reply:
See https://github.com/Snowfork/polkadot-ethereum/pull/314#discussion_r598653157